File: packages/agent-protocol-domain/src/tracking-retention-settings-service-mutation.ts
import { type Infer, Schema, withParser } from '@ocentra-parent/schema-domain/effect';
import {
  AgentCommand,
  AgentEvent,
  AgentProtocolDefaults,
  isAgentProtocolLogText,
  type AgentEventEnvelope,
} from './contracts';

const TrackingRetentionMutationText = Schema.String.pipe(Schema.minLength(1));

export const AgentTrackingRetentionSettingsMutationRequestSchema = withParser(
  Schema.Struct({
    requestId: TrackingRetentionMutationText,
    intentId: TrackingRetentionMutationText,
    settingsKind: TrackingRetentionMutationText,
    writeAction: TrackingRetentionMutationText,
    requestedValue: TrackingRetentionMutationText,
    evidenceReferenceIds: Schema.Array(TrackingRetentionMutationText),
    sourceReadModelProofRefs: Schema.Array(TrackingRetentionMutationText),
    writerBoundaryProofRefs: Schema.Array(TrackingRetentionMutationText),
    auditRefs: Schema.Array(TrackingRetentionMutationText),
  })
);

export const AgentTrackingRetentionSettingsMutationResultSchema = withParser(
  Schema.Struct({
    requestId: TrackingRetentionMutationText,
    mutationId: TrackingRetentionMutationText,
    intentId: TrackingRetentionMutationText,
    settingsKind: TrackingRetentionMutationText,
    writeAction: TrackingRetentionMutationText,
    requestedValue: TrackingRetentionMutationText,
    mutationState: Schema.Literal('accepted', 'rejected'),
    rejectionReason: Schema.Union(TrackingRetentionMutationText, Schema.Null),
    serviceMutationExecuted: Schema.Boolean,
    durablePersistenceClaimed: Schema.Literal(false),
    portalUiClaimed: Schema.Literal(false),
    platformRuntimeClaimed: Schema.Literal(false),
    childDeviceDeliveryClaimed: Schema.Literal(false),
    providerDeliveryClaimed: Schema.Literal(false),
    notificationReceiptClaimed: Schema.Literal(false),
    physicalDeviceClaimed: Schema.Literal(false),
    authorityClaimed: Schema.Literal(false),
    productClaimReady: Schema.Literal(false),
    evidenceReferenceIds: Schema.Array(TrackingRetentionMutationText),
    sourceReadModelProofRefs: Schema.Array(TrackingRetentionMutationText),
    writerBoundaryProofRefs: Schema.Array(TrackingRetentionMutationText),
    auditRefs: Schema.Array(TrackingRetentionMutationText),
  }).pipe(
    Schema.filter(
      (result) =>
        result.evidenceReferenceIds.length > 0 ||
        result.mutationState === 'rejected' ||
        'Tracking retention settings mutation results need evidence refs'
    )
  )
);

export type AgentTrackingRetentionSettingsMutationRequest = Infer<
  typeof AgentTrackingRetentionSettingsMutationRequestSchema
>;
export type AgentTrackingRetentionSettingsMutationResult = Infer<
  typeof AgentTrackingRetentionSettingsMutationResultSchema
>;

export type AgentTrackingRetentionSettingsMutationFailureReason =
  | 'wrong-event'
  | 'missing-json-field'
  | 'invalid-json'
  | 'invalid-payload';

export type AgentTrackingRetentionSettingsMutationParseResult =
  | {
      readonly ok: true;
      readonly value: AgentTrackingRetentionSettingsMutationResult;
    }
  | {
      readonly ok: false;
      readonly reason: AgentTrackingRetentionSettingsMutationFailureReason;
    };

export function createTrackingRetentionSettingsMutationPayload(
  request: AgentTrackingRetentionSettingsMutationRequest
): Record<string, string> {
  return {
    [AgentProtocolDefaults.Field.ActivityTrackingRetentionSettingsMutation]: JSON.stringify(
      AgentTrackingRetentionSettingsMutationRequestSchema.parse(request)
    ),
  };
}

export function trackingRetentionSettingsMutationCommandName(): typeof AgentCommand.ActivityTrackingRetentionSettingsMutate {
  return AgentCommand.ActivityTrackingRetentionSettingsMutate;
}

export function parseTrackingRetentionSettingsMutationEvent(
  event: AgentEventEnvelope
): AgentTrackingRetentionSettingsMutationParseResult {
  if (event.event !== AgentEvent.ActivityTrackingRetentionSettingsMutationReported) {
    return mutationFailure('wrong-event');
  }

  const raw = event.payload[AgentProtocolDefaults.Field.ActivityTrackingRetentionSettingsMutation];
  if (!isAgentProtocolLogText(raw)) {
    return mutationFailure('missing-json-field');
  }

  let decoded: unknown;
  try {
    decoded = JSON.parse(raw);
  } catch {
    return mutationFailure('invalid-json');
  }

  const parsed = AgentTrackingRetentionSettingsMutationResultSchema.safeParse(decoded);
  if (!parsed.success) {
    return mutationFailure('invalid-payload');
  }

  return {
    ok: true,
    value: parsed.data,
  };
}

function mutationFailure(
  reason: AgentTrackingRetentionSettingsMutationFailureReason
): AgentTrackingRetentionSettingsMutationParseResult {
  return {
    ok: false,
    reason,
  };
}

File: crates/agent-protocol/src/tracking_retention_settings_service_mutation.rs
use serde::{Deserialize, Serialize};

pub const TRACKING_RETENTION_SETTINGS_MUTATION_STATE_ACCEPTED: &str = "accepted";
pub const TRACKING_RETENTION_SETTINGS_MUTATION_STATE_REJECTED: &str = "rejected";
pub const TRACKING_RETENTION_SETTINGS_MUTATION_REJECTION_INVALID_REQUEST: &str = "invalid-request";

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrackingRetentionSettingsMutationRequest {
    pub request_id: String,
    pub intent_id: String,
    pub settings_kind: String,
    pub write_action: String,
    pub requested_value: String,
    pub evidence_reference_ids: Vec<String>,
    pub source_read_model_proof_refs: Vec<String>,
    pub writer_boundary_proof_refs: Vec<String>,
    pub audit_refs: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrackingRetentionSettingsMutationResult {
    pub request_id: String,
    pub mutation_id: String,
    pub intent_id: String,
    pub settings_kind: String,
    pub write_action: String,
    pub requested_value: String,
    pub mutation_state: String,
    pub rejection_reason: Option<String>,
    pub service_mutation_executed: bool,
    pub durable_persistence_claimed: bool,
    pub portal_ui_claimed: bool,
    pub platform_runtime_claimed: bool,
    pub child_device_delivery_claimed: bool,
    pub provider_delivery_claimed: bool,
    pub notification_receipt_claimed: bool,
    pub physical_device_claimed: bool,
    pub authority_claimed: bool,
    pub product_claim_ready: bool,
    pub evidence_reference_ids: Vec<String>,
    pub source_read_model_proof_refs: Vec<String>,
    pub writer_boundary_proof_refs: Vec<String>,
    pub audit_refs: Vec<String>,
}
